lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/SharedState.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm
+
+import java.io.{IOException, ObjectInputStream}
+import java.util.concurrent.CountDownLatch
+
+import com.microsoft.ml.spark.lightgbm.dataset.DatasetAggregator
+import com.microsoft.ml.spark.lightgbm.params.TrainParams
+import org.apache.spark.sql.types.StructType
+import org.slf4j.Logger
+
+class SharedState extends Serializable {
+  var mainExecutorWorker: Long = LightGBMUtils.getTaskId()
+
+  @volatile var datasetAggregator: Option[DatasetAggregator] = None
+  def linkDatasetAggregator(columnParams: ColumnParams,
+                            schema: StructType,
+                            trainParams: TrainParams): Unit = {
+    if (datasetAggregator.isEmpty) {
+      this.synchronized {
+        if (datasetAggregator.isEmpty) {
+          datasetAggregator =
+            Some(new DatasetAggregator(columnParams, schema, trainParams.executionParams.useSingleDatasetMode,
+              trainParams.executionParams.chunkSize, trainParams.executionParams.matrixType, this))
+        }
+      }
+    }
+  }
+
+  @volatile var isSparse: Option[Boolean] = None
+  def linkIsSparse(isSparse: Boolean): Unit = {
+    if (this.isSparse.isEmpty) {
+      this.synchronized {
+        if (this.isSparse.isEmpty) {
+          this.isSparse = Some(isSparse)
+        }
+      }
+    }
+  }
+
+  @transient @volatile var arrayProcessedSignal: CountDownLatch = new CountDownLatch(0)
+  def incrementArrayProcessedSignal(log: Logger): Int = {
+    this.synchronized {
+      val count = arrayProcessedSignal.getCount().toInt + 1
+      arrayProcessedSignal = new CountDownLatch(count)
+      log.info(s"Task incrementing ArrayProcessedSignal to $count")
+      count
+    }
+  }
+
+  @transient @volatile var doneSignal: CountDownLatch = new CountDownLatch(0)
+  def incrementDoneSignal(log: Logger): Unit = {
+    this.synchronized {
+      val count = doneSignal.getCount().toInt + 1
+      doneSignal = new CountDownLatch(count)
+      log.info(s"Task incrementing DoneSignal to $count")
+    }
+  }
+
+  @volatile var helperStartSignal: CountDownLatch = new CountDownLatch(1)
+
+
+  def resetSingletonDatasetState(): Unit = {
```

Comment:
after lightgbm has finished training in current executor

Reply:
it looks like you are trying to manage a single shared state across multiple calls to fit. Instead, you can make the shared singleton within the fit method and then just throw it away instead of resetting it and managing it. The shared singleton is almost like a hybrid of a class and an object in this sense

Reply:
done!

Reply:
but if you did this you could remove this method right?

Reply:
removed!